wallet/src/bin/faucet_server.rs

```diff
@@ -0,0 +1,163 @@
+
+//! # The CAPE Faucet
+//!
+
+extern crate cape_wallet;
+
+use async_std::{
+    sync::{Arc, Mutex},
+    task::{spawn, JoinHandle},
+};
+use cape_wallet::{
+    routes::{init_wallet, Wallet},
+    wallet::CapeWalletError,
+    web::default_storage_path,
+};
+use jf_cap::{
+    keys::{UserKeyPair, UserPubKey},
+    structs::AssetCode,
+};
+use rand_chacha::{rand_core::SeedableRng, ChaChaRng};
+use seahorse::loader::Loader;
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use std::path::PathBuf;
+use structopt::StructOpt;
+use tide::StatusCode;
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    name = "CAPE Faucet Server",
+    about = "Grants a native asset seed to a provided UserPubKey"
+)]
+pub struct FaucetOptions {
+    /// mnemonic for the faucet wallet
+    #[structopt(long = "mnemonic", env = "MNEMONIC", default_value = "")]
+    pub mnemonic: String,
+
+    /// path to the faucet wallet
+    #[structopt(long = "faucet_wallet_path", env = "FAUCET_WALLET_PATH")]
+    pub faucet_wallet_path: PathBuf,
+
+    /// password on the faucet account keyfile
+    #[structopt(long = "faucet_password", env = "FAUCET_PASSWORD", default_value = "")]
+    pub faucet_password: String,
+
+    /// binding port for the faucet service
+    #[structopt(long = "faucet_port", env = "FAUCET_PORT", default_value = "50079")]
+    pub faucet_port: String,
+
+    /// size of transfer for faucet grant
+    #[structopt(long = "grant_size", env = "FAUCET_GRANT", default_value = "5000")]
+    pub grant_size: u64,
+
+    /// fee for faucet grant
+    #[structopt(long = "fee_size", env = "FAUCET_FEE", default_value = "100")]
+    pub fee_size: u64,
+}
+
+#[derive(Clone)]
+struct FaucetState {
+    wallet: Arc<Mutex<Wallet>>,
+    grant_size: u64,
+    fee_size: u64,
+}
+
+#[derive(Debug, Snafu, Serialize, Deserialize)]
+#[snafu(module(error))]
+pub enum FaucetError {
+    #[snafu(display("error in faucet transfer: {}", msg))]
+    Transfer { msg: String },
+
+    #[snafu(display("internal server error: {}", msg))]
+    Internal { msg: String },
+}
+
+impl net::Error for FaucetError {
+    fn catch_all(msg: String) -> Self {
+        Self::Internal { msg }
+    }
+    fn status(&self) -> StatusCode {
+        match self {
+            Self::Transfer { .. } => StatusCode::BadRequest,
+            Self::Internal { .. } => StatusCode::InternalServerError,
+        }
+    }
+}
+
+pub fn faucet_server_error<E: Into<FaucetError>>(err: E) -> tide::Error {
+    net::server_error(err)
+}
+
+pub fn faucet_error(source: CapeWalletError) -> tide::Error {
+    faucet_server_error(FaucetError::Transfer {
+        msg: source.to_string(),
+    })
+}
+
+async fn request_fee_assets(
+    mut req: tide::Request<FaucetState>,
+) -> Result<tide::Response, tide::Error> {
+    let bytes = req.body_bytes().await?;
+    let pub_key: UserPubKey = bincode::deserialize(&bytes)?;
```

Comment:
I see we sometimes use `bincode` and sometimes `TaggedBase64`. I'm wondering what the rationale is for using `bincode` here and in the address book.

Reply:
In general, it's more appropriate to use binary (or Json) for POST operations and Base64 for GET (which is embedded in the URL)